### PR TITLE
301 redirect in VirtualHost when server HttpsRedirect true 

### PR DESCRIPTION
**Pull request by zufardhiyaulhaq**
fix https://github.com/istio/istio/issues/34609

Signed-off-by: Zufar Dhiyaulhaq <zufardhiyaulhaq@gmail.com>

**Please provide a description of this PR:**

previously, we only create custom virtual hosts with 301 redirections when there is **no virtual service available** for a specific server in the Gateway object with HttpsRedirect set to true. 

**Gateway server** & **virtual service** will create a **virtual host**. **virtual host** for HTTP might be empty if we create something like this:
```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: my-gateway
  namespace: my-namespace
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    hosts:
    - "example.com"
  - port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - "example.com"
 apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: my-service
  namespace: my-namespace
spec:
  gateways:
  - my-gateway
  hosts:
  - "example.com"
  http:
  - name: example
    match:
    - port: 443
    route:
    - destination:
        host: my-kubernetes-service
        port:
          number: 443
```

but when the HTTP server enabled TLS redirection, this PR will create dummy **virtual host** for HTTP also.
```
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: my-gateway
  namespace: my-namespace
spec:
  selector:
    istio: ingressgateway
  servers:
  - port:
      number: 80
      name: http
      protocol: HTTP
    tls:
      httpsRedirect: true
    hosts:
    - "example.com"
  - port:
      number: 443
      name: https
      protocol: HTTPS
    hosts:
    - "example.com"
 ```
**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 1]
```diff
@@ -389,6 +389,22 @@ func (configgen *ConfigGeneratorImpl) buildGatewayHTTPRouteConfig(node *model.Pr
...
+			if _, exists := vHostDedupMap[host.Name(hostname)]; exists {
+				continue
+			} else if server.Tls != nil && server.Tls.HttpsRedirect {
+				newVHost := &route.VirtualHost{
+					Name:                       domainName(hostname, port),
+					Domains:                    buildGatewayVirtualHostDomains(hostname, port),
+					IncludeRequestAttemptCount: true,
+					RequireTls:                 route.VirtualHost_ALL,
+				}
+				vHostDedupMap[host.Name(hostname)] = newVHost
```
Comment (hzxuzhonghu) [inline pilot/pkg/networking/core/v1alpha3/gateway.go:405]: It seems these are copied from L318-L333, so i guess we could remove the above `if len(virtualServices) == 0 && server.Tls != nil && server.Tls.HttpsRedirect {` branch

Comment (zufardhiyaulhaq) [reply pilot/pkg/networking/core/v1alpha3/gateway.go]: yeah make sense, removing the duplicate code

**Review by howardjohn — COMMENTED** [review]
This seems like it will re-introduce the regression fixed by https://github.com/istio/istio/pull/29895

**Review by howardjohn — CHANGES_REQUESTED** [review]
I don't think this is safe, see comment above.

**Comment by zufardhiyaulhaq** [comment]
> I don't think this is safe, see comment above.

I cannot see your comment.

**Comment by howardjohn** [comment]
> The other case is that a user can have an HTTP server but still "bypass" the httpsRedirect with X-Forward-Proto and num_trusted_hops, so removing the routes breaks this.

If you remove all of the vhosts then this shouldn't work I would think?

**Comment by zufardhiyaulhaq** [comment]
Hi @howardjohn sorry, forgot to change the description, already updated it to not making confusion.

**Review by howardjohn — COMMENTED** [review]
Can you add an integration test? We already have a similar one, we can expand it for this case:
```
tests/integration/pilot/common/routing.go
904-                    opts: echo.CallOptions{
905-                            Count: 1,
906-                            Port: &echo.Port{
907-                                    Protocol: protocol.HTTP,
908-                            },
909:                            Validator: echo.ExpectCode("301"),
```

**Comment by zufardhiyaulhaq** [comment]
@howardjohn create two integration tests, let me know what do you think and what do you expect.

**Comment by howardjohn** [comment]
FWIW https://github.com/istio/istio/issues/30501#issuecomment-903925200 appears to be not fixed by this PR (but it also was broken before the PR). Maybe we can get that one after this lands. Thanks for the PR, will give it another review shortly

File: pilot/pkg/networking/core/v1alpha3/gateway.go [hunk 2]
```diff
@@ -389,6 +366,25 @@ func (configgen *ConfigGeneratorImpl) buildGatewayHTTPRouteConfig(node *model.Pr
...
+			}
+		}
```
Comment (howardjohn) [inline pilot/pkg/networking/core/v1alpha3/gateway.go:387]:
```suggestion
		for _, hostname := range server.Hosts {
			if !server.GetTls().GetHttpsRedirect() {
				continue
			}
			if vHost, exists := vHostDedupMap[host.Name(hostname)]; exists {
				vHost.RequireTls = route.VirtualHost_ALL
				continue
			}
			newVHost := &route.VirtualHost{
				Name:                       domainName(hostname, port),
				Domains:                    buildGatewayVirtualHostDomains(hostname, port),
				IncludeRequestAttemptCount: true,
				RequireTls:                 route.VirtualHost_ALL,
			}
			vHostDedupMap[host.Name(hostname)] = newVHost
		}
	}
```

Feels a little cleaner than duplicating the check multiple times

Comment (howardjohn) [reply pilot/pkg/networking/core/v1alpha3/gateway.go:387]: Not required though, just personal preference

**Review by howardjohn — APPROVED** [review]

File: pilot/pkg/networking/core/v1alpha3/gateway.go (hunk 2 above)
Comment (zufardhiyaulhaq) [reply pilot/pkg/networking/core/v1alpha3/gateway.go:387]: yeah seems cleaner, let me try to commit the changes

**Pull request merged**